### PR TITLE
Fix Prometheus with Multiple Workers

### PR DESCRIPTION
**Pull request by Bentechy66**

### CODE REVIEW

**Review by thebeanogamer — APPROVED** [review]
:shrug:

**Review by jerbob — COMMENTED** [review]
I'm not sure this is the right approach to a solution for this.
I've just done some light googling and this looks promising:
https://github.com/korfuri/django-prometheus/blob/master/documentation/exports.md#exporting-metrics-in-a-wsgi-application-with-multiple-processes-per-processhttps://echorand.me/posts/python-prometheus-monitoring-options/
I'd much rather we stuck with the existing framework's approaches over adding in tech debt that we'll inevitably remove once we figure out how to better utilise django-prometheus.

I will have a look at implementing this tonight if needed, the docs above should give us plenty to experiment with for now.

**Comment by thebeanogamer** [comment]
> I'm not sure this is the right approach to a solution for this.
> I've just done some light googling and this looks promising:
> https://github.com/korfuri/django-prometheus/blob/master/documentation/exports.md#exporting-metrics-in-a-wsgi-application-with-multiple-processes-per-process
> I'd much rather we stuck with the existing framework's approaches over adding in tech debt that we'll inevitably remove once we figure out how to better utilise django-prometheus.
> 
> I will have a look at implementing this tonight if needed, the docs above should give us plenty to experiment with for now.

This looks really messy, exposing on another port already makes routing through Traefik kinda a pain (and they want to do a range for some mad reason), then unlike @Bentechy66's solution we'd still have the issue where statistics are not consistent between Gunicorn and Daphne.

**Comment by thebeanogamer** [comment]
Furthermore, putting the stats into Redis means we can horizontally scale core, whereas keeping it in the context of a worker prevents that.

**Comment by thebeanogamer** [comment]
@Bentechy66 is this cache written to Postgres, or will we lose stats on Redis restarting?

**Comment by jerbob** [comment]
> @Bentechy66 is this cache written to Postgres, or will we lose stats on Redis restarting?
We will lose stats if redis is restarted while core is still running with this approach, yes.

**Comment by thebeanogamer** [comment]
To be fair, I think we still do at the moment, and if we're scraping to prometheus it doesn't matter.

File: src/stats/signals.py [hunk 1]
```diff
@@ -1,68 +1,61 @@
+from django.core.cache import caches
```
Comment (jerbob) [inline src/stats/signals.py:1]: You should be using the default cache here:
```suggestion
from django.core.cache import cache
```

File: src/stats/signals.py [hunk 2]
```diff
@@ -1,68 +1,61 @@
...
+from django.db.models.signals import post_delete
 
 from backend.signals import websocket_disconnect, websocket_connect, team_create, flag_score, register
 from challenge.models import Solve
 from member.models import Member
 from team.models import Team
 
-member_count = Gauge("member_count", "The number of members currently registered")
-member_count.set(Member.objects.count())
+cache = caches["default"]
```
Comment (jerbob) [inline src/stats/signals.py:10]:
```suggestion
```

File: src/stats/views.py [hunk 3]
```diff
@@ -2,8 +2,10 @@
 from datetime import timezone, datetime
 
 from django.contrib.auth import get_user_model
+from django.core.cache import caches
```
Comment (jerbob) [inline src/stats/views.py:5]:
```suggestion
from django.core.cache import cache
```

File: src/stats/views.py [hunk 4]
```diff
@@ -81,6 +91,13 @@ def version(request):
 
 class PrometheusMetricsView(APIView):
     permission_classes = [IsAdminUser]
+    cache = caches["default"]
```
Comment (jerbob) [inline src/stats/views.py:94]:
```suggestion
```

**Review by jerbob — CHANGES_REQUESTED** [review]

**Review by david-cooke — APPROVED** [review]
This is a horrible hack but its more fixed than it is now

**Pull request merged**